lib/pbench/server/database/models/datasets.py

```diff
@@ -895,6 +899,8 @@ def is_key_path(key: str, valid: List[str]) -> bool:
             True if the path is valid, or False
         """
         k = key.lower()
+        if not re.match(Metadata._valid_key_charset, k):
+            return False
         # Check for exact match
         if k in valid:
             return True
```

Comment:
Why/how is the test at line 902 necessary or helpful, given the other tests which follow it?  It seems like you're imposing a lot of overhead on the normal, successful path.

The only case I see of things going wrong is if `k` matches an "open namespace", in which case perhaps line 902 should be combined with line 912 or placed just before it.

Reply:
Yes, I could and probably should have put the character set validation at the end and only bother once we've matched an open namespace pattern. Ah, well.